tuning/tune.py

```diff
@@ -647,65 +716,59 @@ def tune(
     if output is None:
         output = get_default_output_dir()
 
-    # Create the directory if it does not exist    
+    # Create the directory if it does not exist
     makedirs(str(output), exist_ok=True)
 
     tune_logger.debug(f"Output directory {output}")
     tune_logger.debug(f"Processing {input_file}")
     mlir_template = read_input_mlir(input_file)
+    mlir_text = "".join(mlir_template)
 
-    detected_mmt = is_mmt(mlir_template)
-    detected_conv = is_conv(mlir_template)
-    detected_contract = is_contract(mlir_template)
-    detected_batch_matmul = is_batch_matmul(mlir_template)
-    assert [
-        detected_mmt,
-        detected_conv,
-        detected_contract,
-        detected_batch_matmul,
-    ].count(True) == 1
+    mlir_module = parse_mlir(mlir_text)
+    walk_result = walk_mlir_op(mlir_module)
+    assert walk_result.dispatch_kind != None
 
     # Save the input file as the first candidate.
     with open(path.join(output, f"0.mlir"), "w") as f:
-        f.write("".join(mlir_template))
+        f.write(mlir_text)
 
-    if detected_mmt:
-        M, N, K = get_shapes_mmt(mlir_template)
-        tune_logger.debug(f"Matmul shape: [{M}, {N}, {K}]")
+    if walk_result.dispatch_kind == DispatchKind.is_conv:
+        n, oh, ow, oc, fh, fw, ic = get_shapes_conv(mlir_template)
+        tune_logger.debug(f"Conv shape: [n{n}, oh{oh}, oc{oc}, fh{fh}, fw{fw}, ic{ic}]")
+        M = oh * ow
+        N = oc
+        K = fh * fw * ic
+        tune_logger.debug(f"Equivalent matmul shape: [{M}, {N}, {K}]")
```

Comment:
I didn't change anything here, I swapped the conditions from `if mmt elif conv` to `if conv elif mmt`, which would match the same order in dispatch kind enum